cheat-sheet.md

```diff
@@ -0,0 +1,250 @@
+                        9 + 7 = 16
+                                16 + 9 = 25 // done
+
+list.foldRight(0)(_ + _)
+// This is the only valid order
+0 + 9 = 9
+        9 + 7 = 16
+                16 + 5 = 21
+                         21 + 3 = 24
+                                  24 + 1 = 25 // done
+
+list.fold(0)(_ + _) // 25
+// One of the many valid orders
+0 + 1 = 1    0 + 3 = 3             0 + 5 = 5
+        1            3 + 7 = 10            5 + 9 = 14    
+        1                    10          +         14 = 24
+        1                        +                      24 = 25 // done
+
+```
+
+### Algebraic Data Type (ADT)
+ADTs define a fixed set of all possible values of a given type. Two common classes of algebraic types are product types `AND` and sum types `OR`. 
+For sum types, the number of all its possible values is the sum or disjoint union of the number of all values of the two underlying types. 
+The values of a product type typically contain several values, called fields. For products types, the number of its possible values is the product of the fields.
+
+Values of algebraic types are analysed with pattern matching, which identifies a value by its constructor or field names and extracts the data it contains.
+
+### Sealed trait
+When defining an algebraic data type using sealed traits, it allows the compiler to exhaustively check the possible cases in match expressions. The compiler will emit a warning if the match expression isn’t exhaustive. The compiler knows all of the subtypes of the trait that can possibly exist as they can only be extended in the file. 
+
+```
+sealed trait Boolean
+
+case object True extends Boolean
+
+case object False extends Boolean
+```
+
+### Pattern matching
+Pattern matching is a mechanism for checking a value against a pattern. Pattern matching is similar to a series of if/else statements or a switch statement in other languages.
+
+```
+def numToString(num: Int): String =
+  num match {
+    case 0 => "zero"
+    case 1 => "one"
+    case 2 => "two"
+    case _ => "many"
+  }
+```
+
+### map
+Map works by applying a function to each element in the list.
+
+```
+def map[A,B](fa: F[A])(f: A => B): F[B]
+
+def map[A,B](oa: Option[A])(f: A => B): Option[B]
+
+def map[A,B](ea: Either[A])(f: A => B): Either[B]
+
+def map[A,B](ta: Try[A])(f: A => B): Try[B]
+```
+
+###flatMap
+flatMap works by applying a function that returns a sequence for each element in the list, and flattening the results into the original list.
+
+```
+flatMap[A,B](fa: F[A])(f: A => F[B]): F[B]
+
+flatMap[A,B](oa: Option[A])(f: A => Option[B]): Option[B]
+
+flatMap[A,B](ea: Either[A])(f: A => Either[B]): Either[B]
+
+flatMap[A,B](ta: Try[A])(f: A => Try[B]): Try[B]
+```
+
+###For-Comprehensions
+A for-comprehension is syntactic sugar for map, flatMap and filter operations on collections.
+
+The general form is `for {s} yield e`
+`s` is a sequence of generators and filters
+`p <- e` is a generator
+`if f` is a filter
```

Comment:
add an example for the if? or maybe just remove the filtering bit?

Reply:
I think we should remove the filtering bit. Less to explain.